flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java

```diff
@@ -0,0 +1,145 @@
+ * {@link RetryPolicy}. Aimed to curb tail latencies
+ */
+class RetryingExecutor implements AutoCloseable {
+    private static final Logger LOG = LoggerFactory.getLogger(RetryingExecutor.class);
+
+    private final ScheduledExecutorService scheduler;
+
+    RetryingExecutor(int nThreads) {
+        this(SchedulerFactory.create(nThreads, "ChangelogRetryScheduler", LOG));
+    }
+
+    RetryingExecutor(ScheduledExecutorService scheduler) {
+        this.scheduler = scheduler;
+    }
+
+    void execute(RetryPolicy retryPolicy, RunnableWithException action) {
+        LOG.debug("execute with retryPolicy: {}", retryPolicy);
+        RetriableTask task = new RetriableTask(action, retryPolicy, scheduler);
+        scheduler.submit(task);
+    }
+
+    @Override
+    public void close() throws Exception {
+        LOG.debug("close");
+        scheduler.shutdownNow();
+        if (!scheduler.awaitTermination(1, TimeUnit.SECONDS)) {
+            LOG.warn("Unable to cleanly shutdown executorService in 1s");
+        }
+    }
+
+    private static final class RetriableTask implements Runnable {
+        private final RunnableWithException runnable;
+        private final ScheduledExecutorService executorService;
+        private final int current;
+        private final RetryPolicy retryPolicy;
+        private final AtomicBoolean actionCompleted;
+        private final AtomicBoolean attemptCompleted = new AtomicBoolean(false);
+
+        RetriableTask(
+                RunnableWithException runnable,
+                RetryPolicy retryPolicy,
+                ScheduledExecutorService executorService) {
+            this(1, new AtomicBoolean(false), runnable, retryPolicy, executorService);
+        }
+
+        private RetriableTask(
+                int current,
+                AtomicBoolean actionCompleted,
+                RunnableWithException runnable,
+                RetryPolicy retryPolicy,
+                ScheduledExecutorService executorService) {
+            this.current = current;
+            this.runnable = runnable;
+            this.retryPolicy = retryPolicy;
+            this.executorService = executorService;
+            this.actionCompleted = actionCompleted;
+        }
+
+        @Override
+        public void run() {
+            if (!actionCompleted.get()) {
+                Optional<ScheduledFuture<?>> timeoutFuture = scheduleTimeout();
+                try {
+                    runnable.run();
+                    actionCompleted.set(true);
+                    attemptCompleted.set(true);
+                } catch (Exception e) {
+                    handleError(e);
+                } finally {
+                    timeoutFuture.ifPresent(f -> f.cancel(true));
+                }
+            }
+        }
+
+        private void handleError(Exception e) {
+            LOG.trace("execution attempt {} failed: {}", current, e.getMessage());
```

Comment:
debug? info? But I would say almost for certain not just trace?

Reply:
Good piont, I'll change it to `info`.